src/ucs/datastruct/hlist.h

```diff
@@ -266,6 +266,22 @@ ucs_hlist_extract_head(ucs_hlist_head_t *head)
          _elem = ucs_hlist_extract_head_elem(_head, typeof(*(_elem)), _member))
 
 
+/**
+ * Iterate over detached-head list, while removing the head element passes
+ * @a _cond or the list becomes empty.
+ *
+ * @param _elem     Variable to hold the current list element
+ * @param _head     Pointer to list head.
+ * @param _member   List element inside the containing structure.
+ * @param _cond     Condition to test for @a _head element before extract it.
+ */
+#define ucs_hlist_for_each_extract_if(_elem, _head, _member, _cond) \
+    for (_elem = ucs_hlist_head_elem(_head, typeof(*(_elem)), _member); \
+         _elem != UCS_PTR_BYTE_OFFSET(NULL, -ucs_offsetof(typeof(*(_elem)), _member)) && (_cond); \
```

Comment:
1. (_cond) should be checked first
2. can we re implement ucs_hlist_for_each_extract over ucs_hlist_for_each_extract_if where cond=1

Reply:
1. won't work for empty list
2. I think yes, but in case of debugging, this macro will be expanded to more complicated expression

Reply:
1. yes you are right
2. ok